equational_theories/ManuallyProved/Equation3342.lean

```diff
@@ -58,4 +63,106 @@ theorem Equation3342_facts : ∃ (G : Type) (_ : Magma G), Facts G [3342] [3456,
     use (0, 0), (0, 0)
     decide
 
+-- https://leanprover.zulipchat.com/#narrow/channel/458659-Equational/topic/Austin.20pairs/near/482525422
+
+lemma Finite.fn_periodic (G : Type*) [Finite G] (f : G -> G) : ∃ p : ℕ, p > 0 ∧ f^[p] = f^[2*p] := by
+  have Finite.fn_eventually_periodic : ∃ s p : ℕ, p > 0 ∧ f^[s] = f^[s+p] := by
+    obtain ⟨p1, p2, lt, heq⟩ : ∃ p1 p2 : ℕ, p1 < p2 ∧ f^[p1] = f^[p2] := by
+      obtain ⟨p1, p2, ne, heq⟩ := Finite.exists_ne_map_eq_of_infinite (Nat.iterate f ·)
+      rcases le_total p1 p2 with h_le | h_le
+      . refine ⟨p1, p2, Ne.lt_of_le ne h_le, heq⟩
+      . refine ⟨p2, p1, Ne.lt_of_le (Ne.symm ne) h_le, Eq.symm heq⟩
+    let p := p2 - p1
+    have : f^[p1] = f^[p1 + p] := by
+      unfold p
+      rw [← Nat.add_sub_assoc (by linarith)]
+      simp only [heq, add_tsub_cancel_left]
+    refine ⟨p1, p, by simp only [gt_iff_lt, tsub_pos_iff_lt, lt, p], this⟩
+  obtain ⟨s, p, hpgt, hp⟩ := Finite.fn_eventually_periodic
+  have hmod (n j : ℕ) : f^[s + j] = f^[s + j + n*p] := by
+    induction n with
+    | zero => simp only [zero_mul, add_zero]
+    | succ i ih =>
+      have : s + j + (i + 1) * p = s + p + (j + i * p) := by simp_arith only [Nat.succ_mul]
+      rw [this, Function.iterate_add f (s + p), ← hp, ← Function.iterate_add, ← Nat.add_assoc, ih]
+  rcases eq_zero_or_pos s with h | h
+  . simp [h] at hmod
+    have : f^[p] = f^[2*p] := by simp_arith only [hmod 1 p]
+    refine ⟨p, hpgt, this⟩
+  . let n := s * p
+    have : f^[n] = f^[2*n] := by
+      unfold n
+      obtain ⟨ppred, hppred⟩ := Nat.exists_eq_succ_of_ne_zero (by linarith)
+      rw [hppred, Nat.succ_eq_add_one, Nat.mul_succ, Nat.add_comm]
+      have : 2 * (s + s * ppred) = s + s * ppred + s * p := by simp_arith only [hppred,  Nat.mul_succ]
+      rw [this, ← hmod]
+    have ngt : n > 0 := by apply Nat.mul_pos h hpgt
+    refine ⟨n, ngt, this⟩
+
+lemma main_result (G : Type*) [Magma G] [Finite G] (h : Equation3342 G) : Equation3522 G ∧ Equation4118 G := by
```

Comment:
This is a weird decomposition into lemmas, but I can live with it.

Reply:
If you mean that there are two lemmas (fn_periodic and main_result), the first lemma will probably be reused in a future proof so it was easier to keep it separate. If you mean the fact that this lemma proves both equations, I agree that it's weird, but trying to construct it so the equation-specific proofs were in the separate implications led to requiring some of the internal hypotheses to be split out and that led to thers needing to be separate and it seemed worse than this approach in the end, but I'm a Lean-beginner and very open to feedback on how to structure it better!

Reply:
You know better has to structure your proof than I do.

I just know the common pattern — when two proofs share a lot of code, it is better to extract the shared part into a lemma than to merge the two theorems in order to have the shared parts only once. You can also use ˙private lemma˙ if you want.